File: ports/cpp/tests/check_vectors.cpp
// Memeriksa port C++ terhadap test vector dari implementasi TS acuan.
// Exit 0 hanya kalau SEMUA check lulus.
//
// Membaca format datar (.txt) yang diturunkan dari JSON acuan, supaya tidak
// perlu library JSON — header ini sengaja bebas dependensi.

#include "../include/lombok_ecc.hpp"

#include <cstdint>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using lombokecc::GF256;
using lombokecc::ReedSolomon;

static std::size_t g_checks = 0;
static std::size_t g_failed = 0;

static void ok(bool cond, const std::string& label, const std::string& detail = "") {
  ++g_checks;
  if (cond) {
    std::cout << "  PASS  " << label << "\n";
  } else {
    ++g_failed;
    std::cout << "  FAIL  " << label;
    if (!detail.empty()) std::cout << "  — " << detail;
    std::cout << "\n";
  }
}

static std::vector<std::uint8_t> fromHex(const std::string& s) {
  if (s == "-") return {};
  std::vector<std::uint8_t> out;
  out.reserve(s.size() / 2);
  for (std::size_t i = 0; i + 1 < s.size(); i += 2) {
    out.push_back(static_cast<std::uint8_t>(std::stoul(s.substr(i, 2), nullptr, 16)));
  }
  return out;
}

static std::string toHex(const std::vector<std::uint8_t>& b) {
  static const char* d = "0123456789abcdef";
  std::string s;
  s.reserve(b.size() * 2);
  for (std::uint8_t x : b) {
    s.push_back(d[x >> 4]);
    s.push_back(d[x & 0xf]);
  }
  return s;
}

int main(int argc, char** argv) {
  const std::string path = (argc > 1) ? argv[1] : "../vectors/lombok-ecc-vectors-v1.txt";
  std::ifstream in(path);
  if (!in) {
    std::cerr << "Vector tidak ditemukan: " << path << "\n";
    return 2;
  }

  std::vector<std::string> lines;
  for (std::string line; std::getline(in, line);) lines.push_back(line);

  std::size_t n = 255, k = 239, t = 8;
  std::uint32_t poly = 0;
  for (const auto& line : lines) {
    std::istringstream ss(line);
    std::string tag, key;
    ss >> tag;
    if (tag != "PARAM") continue;
    std::size_t v;
    ss >> key >> v;
    if (key == "n") n = v;
    else if (key == "k") k = v;
    else if (key == "t") t = v;
    else if (key == "poly") poly = static_cast<std::uint32_t>(v);
  }

  std::cout << "Port C++ vs vector acuan (commit 313f827)\n";

  std::cout << "\n[1] Parameter\n";
  ReedSolomon rs(n, k);
  ok(rs.n == n, "n cocok");
  ok(rs.k == k, "k cocok");
  ok(rs.t == t, "t cocok");
  ok(lombokecc::kPrimitivePoly == poly, "polinomial primitif cocok");

  GF256 gf;
  std::size_t gfBad = 0, encI = 0, decI = 0, overI = 0, validI = 0;
  int section = 0;

  for (const auto& line : lines) {
    std::istringstream ss(line);
    std::string tag;
    ss >> tag;

    if (tag == "TABLE") {
      if (section < 2) { std::cout << "\n[2] Tabel GF(256)\n"; section = 2; }
      std::string which, hex;
      ss >> which >> hex;
      if (which == "exp") {
        std::vector<std::uint8_t> e(gf.exp.begin(), gf.exp.end());
        ok(toHex(e) == hex, "tabel exp identik (512 entri)");
      } else {
        std::vector<std::uint8_t> logLe;
        logLe.reserve(512);
        for (std::uint16_t v : gf.log) {
          logLe.push_back(static_cast<std::uint8_t>(v & 0xff));
          logLe.push_back(static_cast<std::uint8_t>(v >> 8));
        }
        ok(toHex(logLe) == hex, "tabel log identik (256 x uint16 LE)");
      }
    } else if (tag == "GFOP") {
      if (section < 3) { std::cout << "\n[3] Operasi GF(256)\n"; section = 3; }
      std::string op;
      int a, b, expect;
      ss >> op >> a >> b >> expect;
      int got = -2;
      try {
        if (op == "mul") got = gf.mul(static_cast<std::uint8_t>(a), static_cast<std::uint8_t>(b));
        else if (op == "div") got = gf.div(static_cast<std::uint8_t>(a), static_cast<std::uint8_t>(b));
        else if (op == "inv") got = gf.inv(static_cast<std::uint8_t>(a));
        else if (op == "pow") got = gf.pow(static_cast<std::uint8_t>(a), static_cast<std::size_t>(b));
      } catch (...) { got = -1; }
      if (got != expect) {
        ++gfBad;
        std::cout << "        " << op << "(" << a << "," << b << ") = " << got
                  << ", harusnya " << expect << "\n";
      }
    } else if (tag == "GEN") {
      ok(gfBad == 0, "operasi GF cocok", std::to_string(gfBad) + " menyimpang");
      std::cout << "\n[4] Polinomial generator\n";
      std::size_t degree;
      std::string hex;
      ss >> degree >> hex;
      ok(toHex(rs.g) == hex, "koefisien generator identik");
      ok(rs.g.size() - 1 == degree, "derajat generator cocok");
    } else if (tag == "ENC") {
      if (section < 5) { std::cout << "\n[5] Encode\n"; section = 5; }
      std::size_t len;
      std::string msgHex, cwHex;
      ss >> len >> msgHex >> cwHex;
      try {
        auto cw = rs.encode(fromHex(msgHex));
        ok(toHex(cw) == cwHex,
           "encode vektor " + std::to_string(encI) + " (len " + std::to_string(len) + ") identik");
      } catch (const std::exception& e) {
        ok(false, "encode vektor " + std::to_string(encI), e.what());
      }
      ++encI;
    } else if (tag == "DEC") {
      if (section < 6) { std::cout << "\n[6] Decode dengan error pada posisi tertentu\n"; section = 6; }
      int nErr, nonzero;
      std::string recvHex, decHex;
      ss >> nErr >> nonzero >> recvHex >> decHex;
      auto received = fromHex(recvHex);
      ok(rs.isValid(received) == (nonzero == 0),
         "vektor " + std::to_string(decI) + ": status isValid cocok");
      try {
        auto d = rs.decode(received);
        ok(toHex(d) == decHex, "vektor " + std::to_string(decI) + ": " + std::to_string(nErr) +
                                   " error → message identik");
      } catch (const std::exception& e) {
        ok(false, "vektor " + std::to_string(decI) + ": decode " + std::to_string(nErr) + " error",
           e.what());
      }
      ++decI;
    } else if (tag == "OVER") {
      if (section < 7) {
        std::cout << "\n[7] Di atas kapasitas — wajib gagal, bukan sukses diam-diam\n";
        section = 7;
      }
      int nErr, refThrew;
      std::string recvHex;
      ss >> nErr >> refThrew >> recvHex;
      bool threw = false;
      try { rs.decode(fromHex(recvHex)); } catch (...) { threw = true; }
      if (refThrew) {
        ok(threw, "vektor " + std::to_string(overI) + ": " + std::to_string(nErr) +
                      " error → gagal seperti acuan");
      } else {
        ok(true, "vektor " + std::to_string(overI) + ": acuan tidak melempar, port " +
                     (threw ? "melempar" : "juga tidak"));
      }
      ++overI;
    } else if (tag == "VALID") {
      if (section < 8) { std::cout << "\n[8] isValid\n"; section = 8; }
      int expect;
      std::string cwHex;
      ss >> expect >> cwHex;
      ok(rs.isValid(fromHex(cwHex)) == (expect == 1), "isValid vektor " + std::to_string(validI));
      ++validI;
    }
  }

  // [9] Roundtrip mandiri, LCG deterministik (tanpa <random> supaya hasilnya
  // sama di semua implementasi pustaka standar).
  std::cout << "\n[9] Roundtrip mandiri (bukan dari vector)\n";
  std::uint32_t state = 20260812;
  auto next = [&state]() -> std::uint32_t {
    state = state * 1664525u + 1013904223u;
    return state >> 8;  // buang low bits — periode low bits LCG pendek
  };
  std::size_t rtFail = 0;
  for (int trial = 0; trial < 100; ++trial) {
    const std::size_t len = 1 + (next() % 239);
    std::vector<std::uint8_t> msg(len);
    for (auto& b : msg) b = static_cast<std::uint8_t>(next() & 0xff);
    auto cw = rs.encode(msg);

    const std::size_t nErr = next() % 9;
    std::vector<bool> seen(255, false);
    std::size_t placed = 0;
    while (placed < nErr) {
      const std::size_t p = next() % 255;
      if (!seen[p]) {
        seen[p] = true;
        ++placed;
        const std::uint32_t delta = 1 + (next() % 255);  // 1..255
        cw[p] = static_cast<std::uint8_t>((cw[p] + delta) % 256);
      }
    }

    try {
      auto d = rs.decode(cw);
      bool same = true;
      for (std::size_t i = 0; i < len; ++i) {
        if (d[rs.k - len + i] != msg[i]) { same = false; break; }
      }
      if (!same) ++rtFail;
    } catch (...) {
      ++rtFail;
    }
  }
  ok(rtFail == 0, "100 roundtrip acak (0-8 error) pulih", std::to_string(rtFail) + " gagal");

  std::cout << "\n" << std::string(70, '-') << "\n";
  std::cout << "  " << g_checks << " check, " << g_failed << " gagal\n";
  std::cout << std::string(70, '-') << "\n";
  if (g_failed > 0) {
    std::cout << "  HASIL: GAGAL\n";
    return 1;
  }
  std::cout << "  HASIL: LULUS\n";
  return 0;
}

File: ports/rust/src/bin/check_vectors.rs
//! Memeriksa port Rust terhadap test vector dari implementasi TS acuan.
//! Exit 0 hanya kalau SEMUA check lulus.
//!
//! Membaca format datar (`.txt`) yang diturunkan dari JSON acuan, supaya tidak
//! perlu serde — crate ini sengaja bebas dependensi.

use lombokecc::{Gf256, ReedSolomon, PRIMITIVE_POLY};
use std::env;
use std::fs;
use std::process::exit;

struct Counter {
    checks: usize,
    failed: usize,
}

impl Counter {
    fn ok(&mut self, cond: bool, label: &str, detail: &str) {
        self.checks += 1;
        if cond {
            println!("  PASS  {label}");
        } else {
            self.failed += 1;
            if detail.is_empty() {
                println!("  FAIL  {label}");
            } else {
                println!("  FAIL  {label}  — {detail}");
            }
        }
    }
}

fn from_hex(s: &str) -> Vec<u8> {
    if s == "-" {
        return Vec::new();
    }
    (0..s.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&s[i..i + 2], 16).expect("hex tidak valid"))
        .collect()
}

fn to_hex(b: &[u8]) -> String {
    b.iter().map(|x| format!("{x:02x}")).collect()
}

fn main() {
    let path = env::args()
        .nth(1)
        .unwrap_or_else(|| "../vectors/lombok-ecc-vectors-v1.txt".to_string());
    let raw = match fs::read_to_string(&path) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("Vector tidak ditemukan: {path}: {e}");
            exit(2);
        }
    };

    let mut c = Counter { checks: 0, failed: 0 };
    let mut n = 255usize;
    let mut k = 239usize;
    let mut t = 8usize;
    let mut poly = 0u32;

    // Pass 1: parameter
    for line in raw.lines() {
        let f: Vec<&str> = line.split_whitespace().collect();
        if f.first() == Some(&"PARAM") {
            let v: usize = f[2].parse().unwrap();
            match f[1] {
                "n" => n = v,
                "k" => k = v,
                "t" => t = v,
                "poly" => poly = v as u32,
                _ => {}
            }
        }
    }

    println!("Port Rust vs vector acuan (commit 313f827)");

    println!("\n[1] Parameter");
    let rs = ReedSolomon::new(n, k).expect("parameter tidak valid");
    c.ok(rs.n == n, "n cocok", "");
    c.ok(rs.k == k, "k cocok", "");
    c.ok(rs.t == t, "t cocok", "");
    c.ok(PRIMITIVE_POLY == poly, "polinomial primitif cocok", "");

    let gf = Gf256::new();
    let mut gf_bad = 0usize;
    let mut enc_i = 0usize;
    let mut dec_i = 0usize;
    let mut over_i = 0usize;
    let mut valid_i = 0usize;
    let mut section = 0u8;

    for line in raw.lines() {
        let f: Vec<&str> = line.split_whitespace().collect();
        match f.first() {
            Some(&"TABLE") => {
                if section < 2 {
                    println!("\n[2] Tabel GF(256)");
                    section = 2;
                }
                if f[1] == "exp" {
                    c.ok(to_hex(&gf.exp) == f[2], "tabel exp identik (512 entri)", "");
                } else {
                    let mut log_le = Vec::with_capacity(512);
                    for v in gf.log.iter() {
                        log_le.push((v & 0xff) as u8);
                        log_le.push((v >> 8) as u8);
                    }
                    c.ok(to_hex(&log_le) == f[2], "tabel log identik (256 x uint16 LE)", "");
                }
            }
            Some(&"GFOP") => {
                if section < 3 {
                    println!("\n[3] Operasi GF(256)");
                    section = 3;
                }
                let a: u32 = f[2].parse().unwrap();
                let b: u32 = f[3].parse().unwrap();
                let expect: i32 = f[4].parse().unwrap();
                let got: i32 = match f[1] {
                    "mul" => gf.mul(a as u8, b as u8) as i32,
                    "div" => gf.div(a as u8, b as u8).map(|v| v as i32).unwrap_or(-1),
                    "inv" => gf.inv(a as u8).map(|v| v as i32).unwrap_or(-1),
                    "pow" => gf.pow(a as u8, b as usize) as i32,
                    _ => -2,
                };
                if got != expect {
                    gf_bad += 1;
                    println!("        {}({a},{b}) = {got}, harusnya {expect}", f[1]);
                }
            }
            Some(&"GEN") => {
                c.ok(gf_bad == 0, "operasi GF cocok", &format!("{gf_bad} menyimpang"));
                println!("\n[4] Polinomial generator");
                let degree: usize = f[1].parse().unwrap();
                c.ok(to_hex(&rs.g) == f[2], "koefisien generator identik", "");
                c.ok(rs.g.len() - 1 == degree, "derajat generator cocok", "");
            }
            Some(&"ENC") => {
                if section < 5 {
                    println!("\n[5] Encode");
                    section = 5;
                }
                let msg = from_hex(f[2]);
                match rs.encode(&msg) {
                    Ok(cw) => c.ok(
                        to_hex(&cw) == f[3],
                        &format!("encode vektor {enc_i} (len {}) identik", f[1]),
                        "",
                    ),
                    Err(e) => c.ok(false, &format!("encode vektor {enc_i}"), &e.to_string()),
                }
                enc_i += 1;
            }
            Some(&"DEC") => {
                if section < 6 {
                    println!("\n[6] Decode dengan error pada posisi tertentu");
                    section = 6;
                }
                let nonzero = f[2] == "1";
                let received = from_hex(f[3]);
                c.ok(
                    rs.is_valid(&received) == !nonzero,
                    &format!("vektor {dec_i}: status isValid cocok"),
                    "",
                );
                match rs.decode(&received) {
                    Ok(d) => c.ok(
                        to_hex(&d) == f[4],
                        &format!("vektor {dec_i}: {} error → message identik", f[1]),
                        "",
                    ),
                    Err(e) => c.ok(
                        false,
                        &format!("vektor {dec_i}: decode {} error", f[1]),
                        &e.to_string(),
                    ),
                }
                dec_i += 1;
            }
            Some(&"OVER") => {
                if section < 7 {
                    println!("\n[7] Di atas kapasitas — wajib gagal, bukan sukses diam-diam");
                    section = 7;
                }
                let ref_threw = f[2] == "1";
                let threw = rs.decode(&from_hex(f[3])).is_err();
                if ref_threw {
                    c.ok(
                        threw,
                        &format!("vektor {over_i}: {} error → gagal seperti acuan", f[1]),
                        "",
                    );
                } else {
                    let s = if threw { "melempar" } else { "juga tidak" };
                    c.ok(true, &format!("vektor {over_i}: acuan tidak melempar, port {s}"), "");
                }
                over_i += 1;
            }
            Some(&"VALID") => {
                if section < 8 {
                    println!("\n[8] isValid");
                    section = 8;
                }
                let expect = f[1] == "1";
                c.ok(
                    rs.is_valid(&from_hex(f[2])) == expect,
                    &format!("isValid vektor {valid_i}"),
                    "",
                );
                valid_i += 1;
            }
            _ => {}
        }
    }

    // [9] Roundtrip mandiri dengan LCG deterministik (tanpa dependensi rand).
    println!("\n[9] Roundtrip mandiri (bukan dari vector)");
    let mut state: u32 = 20260812;
    let mut next = || {
        state = state.wrapping_mul(1664525).wrapping_add(1013904223);
        state >> 8 // buang low bits — periode low bits LCG pendek
    };
    let mut rt_fail = 0usize;
    for _ in 0..100 {
        let len = 1 + (next() as usize % 239);
        let msg: Vec<u8> = (0..len).map(|_| (next() & 0xff) as u8).collect();
        let mut cw = rs.encode(&msg).expect("encode gagal");

        let n_err = next() as usize % 9;
        let mut seen = vec![false; 255];
        let mut placed = 0;
        while placed < n_err {
            let p = next() as usize % 255;
            if !seen[p] {
                seen[p] = true;
                placed += 1;
                let delta = 1 + (next() % 255) as u16;
                cw[p] = ((cw[p] as u16 + delta) % 256) as u8; // delta 1..255
            }
        }

        match rs.decode(&cw) {
            Ok(d) => {
                if d[rs.k - len..] != msg[..] {
                    rt_fail += 1;
                }
            }
            Err(_) => rt_fail += 1,
        }
    }
    c.ok(rt_fail == 0, "100 roundtrip acak (0-8 error) pulih", &format!("{rt_fail} gagal"));

    println!("\n{}", "-".repeat(70));
    println!("  {} check, {} gagal", c.checks, c.failed);
    println!("{}", "-".repeat(70));
    if c.failed > 0 {
        println!("  HASIL: GAGAL");
        exit(1);
    }
    println!("  HASIL: LULUS");
}
